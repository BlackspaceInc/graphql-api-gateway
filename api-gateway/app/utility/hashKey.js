import { isString } from "lodash";

import stringify from "safe-stable-stringify";
import { murmurHash128 } from "murmurhash-native";

// A stable, sorted key hashing function so that any deep-equal object
// will match. Currently it is simply a near JSON.stringify for objects,
// which could get large in memory. If this is a concern, we could find a
// fast hash algo to standardize memory consumption.
export function stableStringKey(key) {
  return isString(key) ? key : stringify(key);
}

// We hash the stableStringKey through a consistent, fast hash algorithm.
// We use the 128bit version to ensure collisions are near impossible.
// Function picked by suggestion in this article:
// https://medium.com/@drainingsun/in-search-of-a-good-node-js-hashing-algorithm-8052b6923a3b
export default function hashKey(key) {
  return murmurHash128(stableStringKey(key));
}
