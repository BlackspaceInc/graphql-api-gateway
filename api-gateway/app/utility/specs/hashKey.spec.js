import hashKey, { stableStringKey } from "../hashKey";

describe("hashKey and stableStringKey", () => {
  describe("primatives", () => {
    it("stringifies as is", () => {
      expect(stableStringKey("hey")).toEqual("hey");
    });

    it("stringifies other primatives", () => {
      expect(stableStringKey(false)).toEqual("false");
      expect(stableStringKey(null)).toEqual("null");
      expect(stableStringKey(42)).toEqual("42");
    });

    it("hashes string consistently", () => {
      expect(hashKey("hey")).toMatchSnapshot();
    });

    it("hashes boolean consistently", () => {
      expect(hashKey(false)).toMatchSnapshot();
    });

    it("hashes number consistently", () => {
      expect(hashKey(42)).toMatchSnapshot();
    });

    it("hashes null consistently", () => {
      expect(hashKey(null)).toMatchSnapshot();
    });
  });

  describe("objects", () => {
    const obj = {
      a: "hi",
      b: {
        x: 1,
        y: 2,
      },
      c: [true, false],
    };

    const sameValueObj = {
      c: [true, false],
      a: "hi",
      b: {
        y: 2,
        x: 1,
      },
    };

    const differentObj = {
      ultimately: [true, false],
      a: "hi",
      b: {
        y: 2,
        x: 1,
      },
    };

    it("stringifies objects", () => {
      expect(stableStringKey(obj))
        .toEqual("{\"a\":\"hi\",\"b\":{\"x\":1,\"y\":2},\"c\":[true,false]}");
    });

    it("stringifies to match any deep equal object, regardless of order", () => {
      expect(stableStringKey(obj)).toEqual(stableStringKey(sameValueObj));
    });

    it("hashes the same objects", () => {
      expect(hashKey(obj)).toEqual(hashKey(sameValueObj));
    });

    it("does not match different object hashes", () => {
      expect(hashKey(obj)).not.toEqual(hashKey(differentObj));
    });

    it("consistently hashes object", () => {
      expect(hashKey(obj)).toMatchSnapshot();
    });
  });

  describe("arrays", () => {
    const array = [
      { a: null, b: 3, c: false },
      { happiness: "low", sadness: "very" },
    ];

    const deepEqualArray = [
      { c: false, a: null, b: 3 },
      { sadness: "very", happiness: "low" },
    ];

    const notEqualArray = [
      { sadness: "very", happiness: "low" },
      { c: false, a: null, b: 3 },
    ];

    it("stringifies objects", () => {
      expect(stableStringKey(array))
        .toEqual("[{\"a\":null,\"b\":3,\"c\":false},{\"happiness\":\"low\",\"sadness\":\"very\"}]");
    });

    it("stringifies to match any deep equal array, regardless of order", () => {
      expect(stableStringKey(array)).toEqual(stableStringKey(deepEqualArray));
    });

    it("hashes the same arrays to the same hash", () => {
      expect(hashKey(array)).toEqual(hashKey(deepEqualArray));
    });

    it("does not match different array hashes", () => {
      expect(hashKey(array)).not.toEqual(hashKey(notEqualArray));
    });

    it("consistently hashes array", () => {
      expect(hashKey(array)).toMatchSnapshot();
    });
  });
});
