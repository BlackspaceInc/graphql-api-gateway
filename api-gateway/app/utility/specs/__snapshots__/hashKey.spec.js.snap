// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hashKey and stableStringKey arrays consistently hashes array 1`] = `"3c02c4eb2d3d4ecdf6022b85b981a2b6"`;

exports[`hashKey and stableStringKey objects consistently hashes object 1`] = `"bdbdd01f6f331e66202f530f4299900b"`;

exports[`hashKey and stableStringKey primatives hashes boolean consistently 1`] = `"c4bf6d05e5b795e4ccc3d496069554e8"`;

exports[`hashKey and stableStringKey primatives hashes null consistently 1`] = `"15c50f0697e94e349d6b5e59d5ba43f4"`;

exports[`hashKey and stableStringKey primatives hashes number consistently 1`] = `"b68fda223f324f6c89a6e71bc6804fe3"`;

exports[`hashKey and stableStringKey primatives hashes string consistently 1`] = `"0c2b15608511cf6ab80be99ecc3f7b1d"`;
